src/commands/step_processor.go

```diff
@@ -86,21 +96,109 @@ stderr:
 %v
 `, exitCode, stepID, command, args, stdout, stderr)
 	}
-	s.sendStepReply(stepType, stepID, stdout, stderr, exitCode)
+
+	return s.createStepReply(stepType, stepID, stdout, stderr, exitCode)
 }
 
 func (s *stepSession) handleSteps(steps *models.Steps) {
 	for _, step := range steps.Instructions {
 		if step.Command == "" {
 			errStr := "Missing command"
 			s.Logger().Warn(errStr)
-			s.sendStepReply(step.StepType, step.StepID, "", errStr, -1)
+			s.sendStepReply(s.createStepReply(step.StepType, step.StepID, "", errStr, -1))
 			continue
 		}
-		go s.handleSingleStep(step.StepType, step.StepID, step.Command, step.Args, util.ExecutePrivileged)
+
+		go func(step *models.Step) {
+			if err := s.diagnoseSystem(); err != nil {
+				s.sendStepReply(s.createStepReply(step.StepType, step.StepID, "", err.Error(), -1))
+				return
+			}
+
+			reply := s.handleSingleStep(step.StepType, step.StepID, step.Command, step.Args, util.ExecutePrivileged)
+
+			if err := s.diagnoseSystem(); err != nil {
+				reply.ExitCode = -1
+
+				if reply.Error != "" {
+					reply.Error = fmt.Errorf("%w: %s", err, reply.Error).Error()
+				} else {
+					reply.Error = err.Error()
+				}
+			}
+
+			s.sendStepReply(reply)
+		}(step)
 	}
 }
 
+// diagnoseSystem runs quick validations that need to need to occur before and after each step.
+// This is in order to detect and report known problems otherwise manifest as confusing error messages or stuck the whole system in the steps themselves.
+// One common example of that is virtual media disconnection.
+func (s *stepSession) diagnoseSystem() error {
+	source, err := s.getMountpointSourceDeviceFile()
+
+	if err != nil {
+		log.Warn(err)
+		return nil
+	}
+
+	file, err := os.Open(source)
+	if err != nil {
+		err := errors.Wrap(err, "cannot access the media(ISO) probably media was disconnected")
+		log.Error(err)
+		return err
+	}
+
+	defer file.Close()
+
+	_, err = io.ReadFull(file, make([]byte, 2, 2))
+	if err != nil {
+		err := errors.Wrap(err, "cannot access the media(ISO) probably media was disconnected")
```

Comment:
```suggestion
		err := errors.Wrap(err, "cannot read from the media (ISO) - media was likely disconnected")
```

(Better to have different error messages - in the future we might learn that they are different cases?)